core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java

```diff
@@ -1085,10 +1046,10 @@ public void onPingReceived(final NodesFaultDetection.PingRequest pingRequest) {
                 return;
             }
             logger.debug("got a ping from another master {}. resolving who should rejoin. current ping count: [{}]", pingRequest.masterNode(), pingsWhileMaster.get());
-            clusterService.submitStateUpdateTask("ping from another master", new ClusterStateUpdateTask(Priority.IMMEDIATE) {
+            clusterService.submitStateUpdateTask("ping from another master", new LocalClusterUpdateTask(Priority.IMMEDIATE) {
```

Comment:
hmm.. this is a funky one. It's supposed to run only on masters. On the other hand, I see the complexity involved if we want to call handleAnotherMaster from the normal cluster state processing mechanism as well (where we do need a LocalClusterUpdateTask). We can maybe make rejoin generic, and make `noMaster` and `unchanged` be part of the generic `ClusterTaskResult` but I'm not sure it's worth it. Maybe we should just check we're master within the execute method here and make it a noop if we're not ? wdyt?